atef/check.py

```diff
@@ -0,0 +1,707 @@
+    """Raise this exception to exit a comparator and set severity."""
+    severity = Severity.success
+
+
+class ComparisonError(ComparisonException):
+    """Raise this exception to error out in a comparator."""
+    severity = Severity.error
+
+
+class ComparisonWarning(ComparisonException):
+    """Raise this exception to warn in a comparator."""
+    severity = Severity.warning
+
+
+class ReduceMethod(str, enum.Enum):
+    average = "average"
+    median = "median"
+    sum = "sum"
+    min = "min"
+    max = "max"
+    std = "std"
+
+    @property
+    def method(self) -> Callable:
+        return {
+            ReduceMethod.average: np.average,
+            ReduceMethod.median: np.median,
+            ReduceMethod.sum: np.sum,
+            ReduceMethod.min: np.min,
+            ReduceMethod.max: np.max,
+            ReduceMethod.std: np.std,
+        }[self]
+
+    def reduce(self, values: Sequence[PrimitiveType]) -> PrimitiveType:
+        """
+        Reduce the given values according to the configured method.
+
+        For example, if ``method`` is `ReduceMethod.average`, use `np.average`
+        to reduce the provided values into a scalar result.
+        """
+        return self.method(np.asarray(values))
+
+
+@dataclass(frozen=True)
+class Result:
+    severity: Severity = Severity.success
+    reason: Optional[str] = None
+
+
+def _is_in_range(
+    value: Number, low: Number, high: Number, inclusive: bool = True
+) -> bool:
+    """Is `value` in the range of low to high?"""
+    if inclusive:
+        return low <= value <= high
+    return low < value < high
+
+
+def _raise_for_severity(severity: Severity, reason: str):
+    if severity == Severity.success:
+        return True
+    if severity == Severity.warning:
+        raise ComparisonWarning(reason)
+    raise ComparisonError(reason)
+
+
+success = Result()
+PrimitiveType = Union[str, int, bool, float]
+
+
+@dataclass
+class Value:
+    """A primitive value with optional metadata."""
+    #: The value for comparison.
+    value: PrimitiveType
+    #: A description of what the value represents.
+    description: str = ""
+    #: Relative tolerance value.
+    rtol: Optional[Number] = None
+    #: Absolute tolerance value.
+    atol: Optional[Number] = None
+    #: Severity to set on a match (if applicable).
+    severity: Severity = Severity.success
+
+    def __str__(self) -> str:
+        if self.rtol is not None or self.atol is not None:
+            rtol = f"rtol={self.rtol}" if self.rtol is not None else ""
+            atol = f"atol={self.atol}" if self.atol is not None else ""
+            tolerance = " within " + ", ".join(tol for tol in (rtol, atol) if tol)
+        else:
+            tolerance = ""
+
+        value_desc = f"{self.value}{tolerance} (for a result of {self.severity.name})"
+        if self.description:
+            return f"{self.description} ({value_desc})"
+        return value_desc
+
+    def compare(self, value: PrimitiveType) -> bool:
```

Comment:
Any thoughts on using dunder methods like `__eq__` for things like this?

I'm not sure how I feel. Sometimes `object == 5` feels nice, other times it gets confusing when comparing an object to a primitive.

Reply:
After reading on, it makes sense to keep this as a normal method, unless we wanted to go full confusion by enabling constructs like `value in value_range`

Reply:
An interesting thought, I think - but as your second comment notes, the grammar gets confusing for the non-equality-like variants. It could be named something else like "check" or "run" or...